guardian/recipes-backend
# Re-indexing tool

## What does this change?

- Adds a commandline tool to allow re-indexing of content in the Recipe Backend by grabbing the current published articles from CAPI rather than waiting for them to be re-launched and picked up through Crier in the usual way
- Adds a simple little script to grab the current environment-variable configuration from PROD or CODE environments so that the tool can be easily used
- Remove requirement for node-fetch because we can use built-in `fetch` in Node.js now
- Update tests to mock built-in fetch, also add checks for telemetry calls during take-down that were not previously there
- Add in a telemetry call for a takedown path that was missed

### To Do

## How to test

- Set up your local environment as per the README
- Choose a recipe ID from CODE. (easiest way is to do `curl https://recipes.code.dev-guardianapis.com/index.json | jq .` and then just pick one
- Download the content `curl https://recipes.code.dev-guaredianapis.com/content/{checksum} | jq . > before.json`
- Run `npm run commandline-reindex --recipe-uid {uid}` and watch the logs
- You should see it load up the article, extract all recipes and re-publish.
- Crucially, the SHA value should not change.
- If the SHA does change, download the new version `curl https://recipes.code.dev-guaredianapis.com/content/{checksum} | jq . > after.json` and then check it at https://www.jsondiff.com/

If you're feeling thorough, then run the test again for a valid CAPI path and a valid Composer article ID, as per the README.

**Finally, don't forget**
- structurise and publish a new recipe on CODE, just to prove that it is still publishing properly

## How can we measure success?

Able to push through changes from downstream of Composer more easily

## Have we considered potential risks?

This is not a tool that should be used often.  As long as it's done in accordance with the instructions, there is minimal risk; notes are in the README for higher-risk operations.
